thriftbp/prometheus_test.go

```diff
@@ -100,3 +100,81 @@ func TestPrometheusMetricsMiddleware(t *testing.T) {
 		})
 	}
 }
+
+func TestPrometheusClientMiddleware(t *testing.T) {
+	testCases := []struct {
+		name        string
+		wantErr     bool
+		wantSuccess string
+		exception   string
+	}{
+		{
+			name:        "error",
+			wantErr:     true,
+			wantSuccess: "false",
+			exception:   "thriftbp.PoolError",
+		},
+		{
+			name:        "success",
+			wantErr:     false,
+			wantSuccess: "true",
+			exception:   "",
+		},
+	}
+
+	const (
+		service = "testService"
+		server  = "testServer"
+		method  = "testMethod"
+	)
+
+	for _, tt := range testCases {
+		t.Run(tt.name, func(t *testing.T) {
+			clientLatencyDistribution.Reset()
+			clientRPCRequestCounter.Reset()
+			clientActiveRequests.Reset()
+
+			labelValues := []string{
+				service,
+				method,
+				tt.wantSuccess,
+				tt.exception,
+				"",
+				"",
+				server,
+			}
+
+			requestLabelValues := []string{
+				service,
+				method,
+				server,
+			}
+
+			defer prometheusbp.MetricTest(t, "latency", clientLatencyDistribution).CheckExists()
+			defer prometheusbp.MetricTest(t, "rpc count", clientRPCRequestCounter, labelValues...).CheckDelta(1)
+			defer prometheusbp.MetricTest(t, "active requests", clientActiveRequests, requestLabelValues...).CheckDelta(0)
+
+			client := thrift.WrapClient(mockClient{tt.wantErr}, PrometheusClientMiddleware(service, server))
+			_, err := client.Call(context.Background(), method, nil, nil)
+			if tt.wantErr && err == nil {
+				t.Error("wanted an err, but got nil")
+			}
+			if !tt.wantErr && err != nil {
+				t.Errorf("got err: %v", err)
+			}
+		})
+	}
+}
+
+type mockClient struct {
```

Comment:
I strongly recommend mocking the server and using the real thrift client to connect to an in-process TestBaseplate -- that ensures that you are handling the errors off the wire properly.  For example, the `error` returned by the client when a server returns a particular error will have been serialized and deserialized from the wire, so it might not match exactly what you expect.  A mocked client can potentially return values that a real client will never return, which can lead to bugs with the real code.

Reply:
f.ex https://sourcegraph-staging.build.ue1.snooguts.net/github.snooguts.net/reddit/reddit-service-preference-go/-/blob/service/service_test.go?L333